File: sdk/go/pkce.go
package ggid

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
)

// PKCEPair holds a PKCE code verifier, challenge, and method.
type PKCEPair struct {
	CodeVerifier  string
	CodeChallenge string
	Method        string
}

// GenerateCodeVerifier generates a cryptographically random PKCE code verifier
// per RFC 7636 §4.1 (43-128 chars from unreserved set).
func GenerateCodeVerifier() string {
	b := make([]byte, 64)
	_, _ = rand.Read(b)
	return base64.RawURLEncoding.EncodeToString(b) // 86 chars
}

// GenerateCodeChallenge computes the S256 code challenge from a verifier.
func GenerateCodeChallenge(verifier string) string {
	h := sha256.Sum256([]byte(verifier))
	return base64.RawURLEncoding.EncodeToString(h[:])
}

// GeneratePKCEPair creates a complete PKCE pair with S256 method.
func GeneratePKCEPair() *PKCEPair {
	v := GenerateCodeVerifier()
	return &PKCEPair{
		CodeVerifier:  v,
		CodeChallenge: GenerateCodeChallenge(v),
		Method:        "S256",
	}
}

File: sdk/dart/lib/src/pkce.dart
/// PKCE (Proof Key for Code Exchange) utilities — RFC 7636.
///
/// Since the GGID OAuth server enforces PKCE S256, SDK consumers must generate
/// a code_verifier and code_challenge pair before redirecting to the authorize
/// endpoint, then send the code_verifier with the token exchange request.

library ggid.pkce;

import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';
import 'package:crypto/crypto.dart' as crypto;

/// Generates a cryptographically random code_verifier (43-128 chars, RFC 7636 §4.1).
String generateCodeVerifier() {
  final random = Random.secure();
  final bytes = Uint8List(64);
  for (var i = 0; i < 64; i++) {
    bytes[i] = random.nextInt(256);
  }
  return base64Url.encode(bytes).replaceAll('=', '');
}

/// Derives the S256 code_challenge from a code_verifier (RFC 7636 §4.2).
/// code_challenge = BASE64URL(SHA256(code_verifier))
String generateCodeChallenge(String verifier) {
  final bytes = utf8.encode(verifier);
  final digest = crypto.sha256.convert(bytes);
  return base64Url.encode(digest.bytes).replaceAll('=', '');
}

/// PKCE pair returned by [generatePKCEPair].
class PKCEPair {
  final String codeVerifier;
  final String codeChallenge;
  final String codeChallengeMethod;

  const PKCEPair({
    required this.codeVerifier,
    required this.codeChallenge,
    this.codeChallengeMethod = 'S256',
  });
}

/// Generates a complete PKCE pair for use with the authorize URL.
PKCEPair generatePKCEPair() {
  final verifier = generateCodeVerifier();
  final challenge = generateCodeChallenge(verifier);
  return PKCEPair(
    codeVerifier: verifier,
    codeChallenge: challenge,
    codeChallengeMethod: 'S256',
  );
}